[AutoFill Debugging] Improve debug descriptions for click interactions that only use a location

whsieh opened the pull request:
#### 3a081cf92b798cf55b7db87e407d44b6efd004c9
<pre>
[AutoFill Debugging] Improve debug descriptions for click interactions that only use a location
<a href="https://bugs.webkit.org/show_bug.cgi?id=301584">https://bugs.webkit.org/show_bug.cgi?id=301584</a>
<a href="https://rdar.apple.com/163577331">rdar://163577331</a>

Reviewed by Patrick Angle and Abrar Rahman Protyasha.

Augment descriptions for click interactions that only contain a `location` parameter. Currently, the
descriptions just say `Click at (x, y)`; after these changes, it will say something like:

&gt; `Click at (x, y) on child node of button labeled `“Click me”, with text “Hello”`

...which clearly explains which element is about to be activated.

Test: added to TextExtractionTests.InteractionDebugDescription

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::findNodeAtRootViewLocation):
(WebCore::TextExtraction::dispatchSimulatedClick):
(WebCore::TextExtraction::textDescription):

Pull the main bulk of this logic out into a separate static helper which receives a `Node`, so that
we can reuse it for this new codepath that produces a text description for the hit-tested node in
root view coordinates.

(WebCore::TextExtraction::interactionDescription):
* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::describeTextExtractionInteraction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, InteractionDebugDescription)):

Canonical link: <a href="https://commits.webkit.org/302257@main">https://commits.webkit.org/302257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a36163e15fe80942e409a9f0de5945c0d3ac04e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128512 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135904 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79947 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ca40a7b7-a005-408a-be0d-ba1a9fb3feca) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130384 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/653 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97832 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5ad2470c-ede7-42bb-8429-0a0f0cd1fd02) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115143 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78444 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/df547f25-a1b1-464b-807f-03271101776c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/473 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/33250 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79187 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108895 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33733 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138353 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/616 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/583 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106368 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/660 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106182 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27051 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/516 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30023 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52953 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/673 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63873 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/557 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/627 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1511,13 +1538,12 @@ InteractionDescription interactionDescription(const Interaction& interaction)
+    auto appendElementString = [&](auto&&... args) {
+        auto elementString = textDescription(std::forward<decltype(args)>(args)...);
+        if (elementString.isEmpty())
+            return;
```
Nit: the style guide prefers an explicit template argument list when the parameter type is required in the function body.

```suggestion
    auto appendElementString = [&]<typename... T>(T&&... args) {
        auto elementString = textDescription(std::forward<T>(args)...);
        if (elementString.isEmpty())
            return;
```

whsieh replied:
TIL — thanks for the reviews!

whsieh committed:
[AutoFill Debugging] Improve debug descriptions for click interactions that only use a location
https://bugs.webkit.org/show_bug.cgi?id=301584
rdar://163577331

Reviewed by Patrick Angle and Abrar Rahman Protyasha.

Augment descriptions for click interactions that only contain a `location` parameter. Currently, the
descriptions just say `Click at (x, y)`; after these changes, it will say something like:

> `Click at (x, y) on child node of button labeled `“Click me”, with text “Hello”`

...which clearly explains which element is about to be activated.

Test: added to TextExtractionTests.InteractionDebugDescription

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::findNodeAtRootViewLocation):
(WebCore::TextExtraction::dispatchSimulatedClick):
(WebCore::TextExtraction::textDescription):

Pull the main bulk of this logic out into a separate static helper which receives a `Node`, so that
we can reuse it for this new codepath that produces a text description for the hit-tested node in
root view coordinates.

(WebCore::TextExtraction::interactionDescription):
* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::describeTextExtractionInteraction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/TextExtractionTests.mm:
(TestWebKitAPI::TEST(TextExtractionTests, InteractionDebugDescription)):

Canonical link: https://commits.webkit.org/302257@main
